rascal-lsp/src/main/rascal/library/util/LanguageServer.rsc

```diff
@@ -268,8 +268,19 @@ data LanguageService
     | references    (set[loc] (Focus _focus) referencesService)
     | implementation(set[loc] (Focus _focus) implementationService)
     | codeAction    (list[CodeAction] (Focus _focus) codeActionService)
+    | formatting    (str (Tree _input, FormattingOptions _opts) formattingService)
     ;
 
+data FormattingOptions
+    // If LSP adds more options, add them as keyword arguments for backward compatibility
```

Comment:
This means that the current options are (and will remain) positional parameters, whereas future extensions will become keyword parameters. As ADTs are extensible, have you considered distinct constructors for the various formatting options and passing a `list` of options to the `formattingService`?

Reply:
Thanks, I keep forgetting how a generator can be used to pattern match on single elements.